### PR TITLE
Only animate the  blocks that changed index

### LINKED ISSUE #17383: Safari: Moving Blocks is SUPER slow
**Describe the bug**

**To reproduce**
0. Use Safari.
1. Create a few blocks or open the demo content.
2. Use the block movers.
3. The block(s) move very slowly to the target position. This is even more annoying when you want to move the block(s) several positions up or down.

**Expected behavior**

Moving blocks should be (near) instant.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Mac
 - Browser: Safari
 - Version: 12.1.1 (14607.2.6.1.1)

### PR DESCRIPTION
**Pull request by youknowriad**
closes #17383 

This PR tries to fix the animation performance issues we're seeing in Safari. The idea is to only animate the block if its index changes while previously all blocks were rerendered each time the block list was updated.

Right now, it's not perfect though because for some reason the selected block is not being animated. Need to figure out why.

### CODE REVIEW

**Review by ellatrix — APPROVED** [review]
Perfect.

**Comment by youknowriad** [comment]
I can't explain why the current block is not being animated 🤔

**Comment by youknowriad** [comment]
It looks like the `PreserveScrollInReorder` component is impacting the behavior. I think  `getBoundingClientRect` is being impacted by the page scrolling. I need something to get the absolute position that is stable when you scroll the page.

**Comment by youknowriad** [comment]
@jasmussen @mtias 

So it looks like there's multiple approaches here:

 - master behavior: the selected block is not moving, but all the other blocks  (including the blocks that don't move) are actually being animated to give that impression of movement. (I think we can't continue with this approach due to the performance degradation)

 - current behavior in the PR: the selected block doesn't really get animated, it looks a bit abrupt. The unselected blocks that need to move are being animated but the selected block is not.

 - approach I have tried locally: animate the selected block as well but this will produce a feeling of jumpiness, because as soon as the scrolling happens, it can seem that the block moved up and then is animated down. In reality the block only moved once from but that's not what people will think. To solve this jumpiness of the "preserve scrolling" behavior, I also tried animating the scrolling ("smooth" scrolling), the jumpiness is less visible but it's still there, the only way to avoid the jumpiness could be to have the exact oposite animation (same speed and acceleration) for the scrolling, this won't be easy to achieve.

Thoughts?

**Comment by jasmussen** [comment]
I think there's a near term fix we need, and a longer term. So on a meta level, which ever looks decent and is the fastest one that you're most confident in, is probably what we should go with.

Longer term, we could look at improving it further. Did you try the idea where you only animate the selected block and then 10 before and after it, but use the behavior that is in master? The problem is mainly on REALLY long posts, and those can't fit in a viewport anyway, so there's no need to animate them.

**Comment by youknowriad** [comment]
> Longer term, we could look at improving it further. Did you try the idea where you only animate the selected block and then 10 before and after it,

Well, this PR only animates the selected block and the one that is after or before (the one that needs . to move) so it's even better and for long posts we already disable block animation for posts with more than 200 blocks.

I need help figuring out the best near term solution. What do you think of the current PR state?

**Comment by jasmussen** [comment]
> What do you think of the current PR state?

Here's what I see in Chrome:

chrome

Safari:

saf

Both of those are _ok_, not _great_. But more than great, they are consistent, and it fixes a rather burning issue on the performance front, AND it scales well. 

Here's master:

master

Master is the gold standard. Smooth and lovely. 

It can be hard to capture the difference between those two. But what happens is that only the immediate next block gets animated, the rest "stutter into place". 

To some extent this makes it seem like you're only actually animating the block whose place you're taking with the selected block. 

In other words, it seems like if we DO want the gold standard animation we HAVE to animate all the _visible_ blocks. Can we ever do that? We'd essentially establish a "max blocks animating at one point" so there should be a roof over the performance hit, it wouldn't scale endlessly.

**Comment by jasmussen** [comment]
To elaborate a bit on what I mean by animating the visible blocks. 

- Master is animating `blockClientIds`, ALL blocks.
- This branch is animating `index`, only the selected block.
- Could you animate `index`, plus the 10 blocks from `blockClientIds` before `index`, and the 10 blocks after?

**Comment by youknowriad** [comment]
This branch is not animating the selected block. `index`  is animating all the blocks whose index (position) change in the list. 

To animate +-10blocks we need a selector that changes its value everytime a block get moved within the range of 10 blocks from the current block, it is very hard to achieve, and running that selector for all blocks synchronously is a performance issue on its own.

**Comment by jasmussen** [comment]
Understood, thanks for explaining.

I do think this PR solves the issue and can go in. But at the same time with just one block animating like this, it does feel jumpy to the point that I'm not sure what value it brings over having no animation at all. This is not a strong opinion, I don't mean it to be negative. But it could be a reason to merge a "no animation" PR for the time being, and regroup to see how we can revisit this maybe? Curious what @mtias thinks.

**Comment by youknowriad** [comment]
you also try this branch `try/animating-scrolling-fix`  where I explore the third approach highlighted above, you'll see both the block and the scrolling being animated causing a weird effect.

**Comment by jasmussen** [comment]
> you also try this branch try/animating-scrolling-fix 

Here's what I see:

scroll animate

That seems to be much the same as the other implementation, in that it only ever animates the block whose place is being taken by the block being moved. Which again, is fine, but not sure it's that much better than no animation. It does feel ever so slightly faster, though?

**Comment by youknowriad** [comment]
Can you check now, I believe it's better.

**Comment by jasmussen** [comment]
Fast work! First the good news:

animation

This feels really really good. If this fixes the perf issue, we are definitely on the right path.

The bad news though, I'm seeing some glitchiness around _selecting blocks causing animation to happen_, and subsequently invoking multi select:

glitchiness

The GIF best describes what's up, but essentially select one paragraph, sometimes just doing that causes some animation to happen. 

Then click the paragraph below and sometimes that invokes multi select. 

This does not appear for me in master, so it seems specific to this branch.

**Comment by youknowriad** [comment]
@jasmussen Thanks for testing, I think it's fixed.

**Comment by jasmussen** [comment]
Whatever you pushed right now is SWEET. 

sweet

SHIP IT!
